core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -420,4 +425,113 @@ protected WriteReplicaResult shardOperationOnReplica(BulkShardRequest request, I
         return next;
     }
 
+    public <Request extends ReplicatedWriteRequest<Request>, Response extends ReplicationResponse & WriteResponse>
+    WritePrimaryResult<Request, Response> executeSingleItemBulkRequestOnPrimary(
```

Comment:
I have moved these BWC method to static methods to `SingleWriteOperationUtility`, (suggestion for better name welcomed). The only downside is now `shardOperationOnPrimary` and `shardOperationOnReplica` for bulk shard action is made public, as they are called from transport index and delete actions